### PR TITLE
REF/TST: Fix remaining DatetimeArray with DateOffset arithmetic ops

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Travis fail is the test_rank segfault

File: pandas/core/arrays/timedeltas.py [hunk 1]
```diff
@@ -133,15 +133,31 @@ def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
...
+        verify_integrity = True
 
         freq, freq_infer = dtl.maybe_infer_freq(freq)
 
-        values = np.array(values, copy=False)
-        if values.dtype == np.object_:
-            values = array_to_timedelta64(values)
+        values, inferred_freq = sequence_to_td64ns(values, copy=copy,
+                                                   unit=None)
+        if inferred_freq is not None:
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: nitpick but I would write this as
```
values, inferred_freq = sequence_to_td64ns(
        values, copy=copy, unit=None)
```

File: pandas/core/arrays/timedeltas.py [hunk 2]
```diff
@@ -133,15 +133,31 @@ def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
...
+                                 .format(inferred=inferred_freq,
+                                         passed=freq.freqstr))
+            elif freq_infer:
+                freq = inferred_freq
+                freq_infer = False
+            verify_integrity = False
 
         result = cls._simple_new(values, freq=freq)
+        # check that we are matching freqs
+        if verify_integrity and len(result) > 0:
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: can't you use freq_infer rather than verify_integrity here, seems kind of clunky

File: pandas/tests/arithmetic/test_datetime64.py [hunk 3]
```diff
@@ -1058,9 +1058,232 @@ def test_dti_add_tick_tzaware(self, tz_aware_fixture, box_with_array):
...
+            # If cls_name param is a tuple, then 2nd entry is kwargs for
+            # the offset constructor
+            cls_name, kwargs = cls_and_kwargs
+        else:
+            cls_name = cls_and_kwargs
+            kwargs = {}
+
+        offset_cls = getattr(pd.offsets, cls_name)
+
+        with warnings.catch_warnings(record=True):
```
Comment (jreback) [inline pandas/tests/arithmetic/test_datetime64.py]: this is a monster test, see if you can split a bit

File: pandas/tseries/offsets.py [hunk 4]
```diff
@@ -1328,20 +1346,24 @@ def _end_apply_index(self, dtindex):
 
         base, mult = libfrequencies.get_freq_code(self.freqstr)
         base_period = dtindex.to_period(base)
+        if not isinstance(base_period._data, np.ndarray):
```
Comment (jreback) [inline pandas/tseries/offsets.py]: you seem to be doing this unwrappng a lot, wouldn't it be better to have a ``._to_unwrapped_period(..)``? that just does this (basically to_period() with unwrapping to the base data)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/timedeltas.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: sure

File: pandas/core/arrays/timedeltas.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: ill double check, but probably.  This is a holdover from the Index behavior, where verify_integrity is also a kwarg for `__new__` (which probably should be deprecated)

File: pandas/tseries/offsets.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tseries/offsets.py]: The issue here is that we want to unwrap a PeriodIndex, but leave a PeriodArray unchanged.  Or are you suggesting that `_to_unwrapped_period()` would return `self` for the PeriodArray?  I'm kind of expecting that down the road we'll have some general version of this to allow us to treat EAs and EA-backed Indexes symmetrically.

File: pandas/core/arrays/timedeltas.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: Turns out we can't use freq_infer, but we can use `inferred_freq is None`

File: pandas/tests/arithmetic/test_datetime64.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tests/arithmetic/test_datetime64.py]: sure

File: pandas/tseries/offsets.py (hunk 4 above)
Comment (jreback) [reply pandas/tseries/offsets.py]: just a thought. happy to defer, but i think it does make sense to have methods that unwrap on Index but return self for Array

File: pandas/tests/arithmetic/test_datetime64.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tests/arithmetic/test_datetime64.py]: pushed one more level of indentation into parametrize.  Note a huge difference, but its something.

File: pandas/core/arrays/timedeltas.py [hunk 5]
```diff
@@ -227,6 +242,18 @@ def _validate_fill_value(self, fill_value):
                              "Got '{got}'.".format(got=fill_value))
         return fill_value
 
+    @property
+    def is_monotonic_increasing(self):
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/timedeltas.py]: Where is this used in this PR?

File: pandas/core/arrays/timedeltas.py [hunk 6]
```diff
@@ -162,15 +162,30 @@ def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
...
         return result
 
-    def __new__(cls, values, freq=None, dtype=_TD_DTYPE):
+    def __new__(cls, values, freq=None, dtype=_TD_DTYPE, copy=False):
 
         freq, freq_infer = dtl.maybe_infer_freq(freq)
 
-        values = np.array(values, copy=False)
-        if values.dtype == np.object_:
-            values = array_to_timedelta64(values)
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/timedeltas.py]: What is the more or less the difference between `array_to_timedelta64` and `sequence_to_td64ns` ?

File: pandas/core/arrays/timedeltas.py [hunk 7]
```diff
@@ -162,15 +162,30 @@ def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
...
+            values, copy=copy, unit=None)
+        if inferred_freq is not None:
+            if freq is not None and freq != inferred_freq:
+                raise ValueError('Inferred frequency {inferred} from passed '
+                                 'values does not conform to passed frequency '
+                                 '{passed}'
+                                 .format(inferred=inferred_freq,
+                                         passed=freq.freqstr))
+            elif freq is None:
+                # TODO: should this be the stronger condition `if freq_infer`?
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/timedeltas.py]: Can you clarify this TODO comment?

File: pandas/core/arrays/timedeltas.py [hunk 8]
```diff
@@ -162,15 +162,30 @@ def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
...
+                freq = inferred_freq
+                freq_infer = False
 
         result = cls._simple_new(values, freq=freq)
+        # check that we are matching freqs
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: I think you could simplify a lot of these checks if also pass ``inferred_freq`` (optionally to ``._validate_frequencey``, and for example handle the fact that result could be a scalar (just ignore the validation) / sequence (do the validation)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/timedeltas.py (hunk 8 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: ill take a look at this

File: pandas/core/arrays/timedeltas.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: Sure.  We might also be able to resolve the answer as "no" in this thread.

ATM if we get an `inferred_freq` here, we are setting `freq = inferred_freq` regardless of whether the user passed `freq="infer"`.  In particular, we are ignoring the possibility that the user passed `freq=None` specifically wanting the result to have no freq.

File: pandas/core/arrays/timedeltas.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: `sequence_to_td64ns` is the more general case, wraps `array_to_timedelta64`

File: pandas/core/arrays/timedeltas.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: When calling the constructor with `freq="infer"` we end up calling `tseries.frequencies.infer_freq`, which accesses this property

Comment (jorisvandenbossche) [reply pandas/core/arrays/timedeltas.py]: I think I already expressed this previously, but I would prefer that we have this discussion for all the Arrays, not just for TimedeltaArray (or DatetimeArray), as this is not a datetime-specific attribute. If you don't want to have that discussion first, you can always make it a private attribute here, and check that in `infer_freq` (or call the `algos` methods there if the passed values don't have such an attribute)

Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: If this is a discussion you'd like to have for EAs more generally, go ahead and open an issue for it.

I would be +1 on putting these attributes in the mixin class so that they are available on all three of DTA/TDA/PA, but for now they are only needed on TDA.

Special-casing inside `infer_freq` (or more specifically `FrequencyInferer` is a pretty ugly hack that I'd rather avoid.  That said, if you're going to insist on it, I might as well get it over with.

**Comment by jbrockmendel** [comment]
gentle ping

File: pandas/core/arrays/timedeltas.py [hunk 9]
```diff
@@ -162,15 +162,32 @@ def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
...
+            if freq is not None and freq != inferred_freq:
+                raise ValueError('Inferred frequency {inferred} from passed '
+                                 'values does not conform to passed frequency '
+                                 '{passed}'
+                                 .format(inferred=inferred_freq,
+                                         passed=freq.freqstr))
+            elif freq is None:
+                # TODO: should this be the stronger condition `if freq_infer`?
+                #  i.e what if the user passed `freq=None` and specifically
+                #  wanted freq=None in the result?
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: I am not sure this is a valid condition.  what are you trying to do here?

File: pandas/core/arrays/timedeltas.py [hunk 10]
```diff
@@ -227,6 +244,18 @@ def _validate_fill_value(self, fill_value):
...
+    def is_monotonic_increasing(self):
+        return algos.is_monotonic(self.asi8, timelike=True)[0]
+
+    @property
+    def is_monotonic_decreasing(self):
+        return algos.is_monotonic(self.asi8, timelike=True)[1]
+
+    @property
+    def is_unique(self):
+        return len(unique1d(self.asi8)) == len(self)
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: is this *only* for TimedeltaArray? the index should not be doing this

File: pandas/core/indexes/timedeltas.py [hunk 11]
```diff
@@ -227,6 +227,10 @@ def _format_native_types(self, na_rep=u'NaT', date_format=None, **kwargs):
     # -------------------------------------------------------------------
     # Wrapping TimedeltaArray
 
+    is_monotonic_increasing = Index.is_monotonic_increasing
+    is_monotonic_decreasing = Index.is_monotonic_decreasing
+    is_unique = Index.is_unique
```
Comment (jreback) [inline pandas/core/indexes/timedeltas.py]: you need to time this, we have a dedicated routine in the cython engine for this, the main reason for using it is it a O(n) op once the hashtable is computed, rather than a non-cached computation which you did above.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/timedeltas.py (hunk 11 above)
Comment (jreback) [reply pandas/core/indexes/timedeltas.py]: this is not obvious at all. i would expect a comment on this. maybe even do this in a separate PR, with testing for this. I am not sure its relevant to the changes in this PR.

File: pandas/core/arrays/timedeltas.py (hunk 10 above)
Comment (jreback) [reply pandas/core/arrays/timedeltas.py]: I much prefer comments in the code to non-obvious things.

File: pandas/core/arrays/timedeltas.py (hunk 9 above)
Comment (jreback) [reply pandas/core/arrays/timedeltas.py]: I doubt a user setting a freq is actually used anywhere, not even sure of the purpose of it; nor is it likely tested.

This is just adding complextiy.

File: pandas/core/arrays/timedeltas.py (hunk 10 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: I'll add comments to that effect

File: pandas/core/arrays/timedeltas.py (hunk 9 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: OK, I'll remove the comment

File: pandas/core/indexes/timedeltas.py (hunk 11 above)
Comment (jbrockmendel) [reply pandas/core/indexes/timedeltas.py]: I'll add a comment.  It is definitely relevant to this PR, since without implementing these in the TDA calls to `infer_freq` (via `__new__`) raise `AttributeError`

File: pandas/core/arrays/timedeltas.py (hunk 5 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/timedeltas.py]: @jbrockmendel I would really prefer if you leave this out of this PR (I mean adding the attributes, so which means special casing this in `infer_freq`). 

Looking at the code again, I think the easiest to do is to ensure what is passed to `_FrequencyInferrer` is an actual index. So if we have an array, simply convert to Index without copy.

File: pandas/tests/arithmetic/test_datetime64.py [hunk 12]
```diff
@@ -1058,9 +1058,234 @@ def test_dti_add_tick_tzaware(self, tz_aware_fixture, box_with_array):
...
+                             Timestamp('2000-02-29'),
+                             Timestamp('2000-12-31'),
+                             Timestamp('2000-05-15'),
+                             Timestamp('2001-06-15')])
+        vec = tm.box_expected(vec, box_with_array)
+        vec_items = vec.squeeze() if box_with_array is pd.DataFrame else vec
+
+        offset_cls = getattr(pd.offsets, cls_name)
+
+        with warnings.catch_warnings(record=True):
```
Comment (jorisvandenbossche) [inline pandas/tests/arithmetic/test_datetime64.py]: I don't think the `record=True` is needed? (as you are not checking the captured warnings)

File: pandas/tseries/offsets.py [hunk 13]
```diff
@@ -275,14 +275,17 @@ def apply_index(self, i):
...
-                i = i._shallow_copy(shifted)
+                i = type(i)(shifted, freq=i.freq, dtype=i.dtype)
 
             weeks = (kwds.get('weeks', 0)) * self.n
             if weeks:
                 # integer addition on PeriodIndex is deprecated,
                 #   so we directly use _time_shift instead
                 asper = i.to_period('W')
-                shifted = asper._data._time_shift(weeks)
+                if not isinstance(asper._data, np.ndarray):
```
Comment (jorisvandenbossche) [inline pandas/tseries/offsets.py]: when is this not the case?

(or why was this change needed?)

File: pandas/tseries/offsets.py [hunk 14]
```diff
@@ -912,7 +919,9 @@ def apply(self, other):
     @apply_index_wraps
     def apply_index(self, i):
         shifted = liboffsets.shift_months(i.asi8, self.n, self._day_opt)
-        return i._shallow_copy(shifted)
+        # TODO: going through __new__ raises on call to _validate_frequency;
+        #  are we passing incorrect freq?
```
Comment (jorisvandenbossche) [inline pandas/tseries/offsets.py]: Does this todo needs to be discussed / handled now?

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/timedeltas.py [hunk 15]
```diff
@@ -162,15 +162,29 @@ def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
...
         freq, freq_infer = dtl.maybe_infer_freq(freq)
 
-        values = np.array(values, copy=False)
-        if values.dtype == np.object_:
-            values = array_to_timedelta64(values)
+        values, inferred_freq = sequence_to_td64ns(
+            values, copy=copy, unit=None)
+        if inferred_freq is not None:
+            if freq is not None and freq != inferred_freq:
+                raise ValueError('Inferred frequency {inferred} from passed '
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/timedeltas.py]: This is not covered according to codecov

File: pandas/tseries/offsets.py (hunk 13 above)
Comment (jorisvandenbossche) [reply pandas/tseries/offsets.py]: and something else, this full `if weeks` block does not seem to be covered (don't know if it is a false positive or not)

File: pandas/core/arrays/timedeltas.py (hunk 15 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: Thanks.  I'm fairly sure the matching line in `TimedeltaIndex.__new__` is covered by a targeted test.  I'll update that test to cover this too (and after #23675 I plan to do a pass to de-duplicate a bunch of the DTA/TDA/DTI/TDI constructors)

File: pandas/core/arrays/timedeltas.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: converting to an Index would break the "Array should be ignorant of Index" rule discussed elsewhere.

I'll special-case this to get this over with, but I maintain this is introducing a code smell.

File: pandas/tseries/offsets.py (hunk 13 above)
Comment (jbrockmendel) [reply pandas/tseries/offsets.py]: `asper` at this point may be either a `PeriodIndex` or a `PeriodArray`.  If it is a `PeriodIndex` we want to unwrap it to get a PeriodArray.  If it is a PeriodArray, we _dont_ want to unwrap it since that would give a ndarray

Comment (jbrockmendel) [reply pandas/tseries/offsets.py]: Good note on the `if weeks`.  Adding coverage is out of scope, but I'll add a note in the offsets TODO list

File: pandas/tseries/offsets.py (hunk 14 above)
Comment (jbrockmendel) [reply pandas/tseries/offsets.py]: I'm happy thinking of this as a note to future-me.  Really looking forward to getting the current crop of PRs done with and DatetimeArray out the door.

**Comment by jbrockmendel** [comment]
@jorisvandenbossche I think your main concern has been addressed.  LMK if any others are deal-breakers.

File: pandas/tseries/frequencies.py [hunk 16]
```diff
@@ -295,8 +295,19 @@ def __init__(self, index, warn=True):
         if len(index) < 3:
             raise ValueError('Need at least 3 dates to infer frequency')
 
-        self.is_monotonic = (self.index.is_monotonic_increasing or
-                             self.index.is_monotonic_decreasing)
+        if not hasattr(index, "is_monotonic_increasing"):
```
Comment (jreback) [inline pandas/tseries/frequencies.py]: why don't you define a method on both TimedeltaArray & TimedeltaIndex which basically does the correct thing here?  I find this not very obvious what you are doing and the differencde betwee TDI and TA is pretty stark here (of the calling conventions and such).

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tseries/frequencies.py (hunk 16 above)
Comment (jbrockmendel) [reply pandas/tseries/frequencies.py]: That's what I did originally, but @jorisvandenbossche insisted on not-that.

Comment (jorisvandenbossche) [reply pandas/tseries/frequencies.py]: >  Its a private method, not sure the reason. 

Not sure what you mean with this comment. This is a private method yes, but why would that be a reason that this is not the good place to put it? 
For now, it is also the only place where this logic is needed, so I don't think it is necessarily a bad place. Yes, if this would start to be used in a lot of places, putting it as a method on the Array class makes sense, for now I don't really care. 

What I asked was to just leave out the addition of *public* attributes on the Array classes out of this PR (as that needs a more general discussion). Whether the logic is put here, in a private attribute on the Array, or by doing a `self.index = pd.Index(self.index, copy=False)` to ensure it has the needed properties (and the rest of the code here can stay the same), I don't care much.

Comment (jreback) [reply pandas/tseries/frequencies.py]: because you are putting logic that clearly has a different implementation outside of the class which defines it ; this is breaking all kinds of abstraction here

the inference engine needs to know some info which is better off donenall together for perf reasons - so call a method - that is what they are for 

@jbrockmendel move this to a private method on ta and tdi

Comment (jbrockmendel) [reply pandas/tseries/frequencies.py]: done.  Now frequencies.infer_freq is accessing private attrs, but we can revisit that smell later

**Comment by jbrockmendel** [comment]
Should be all squared away now.

**Comment by jbrockmendel** [comment]
gentle ping.  This and #23829 are the main things standing in the way of declaring DTA/TDA "finished" (internally)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
couple of notes for followup, but lgtm.

@jorisvandenbossche

**Comment by jorisvandenbossche** [comment]
Thanks @jbrockmendel !

**Comment by jbrockmendel** [comment]
Booyah.  Thanks @jorisvandenbossche, this one is a big win.
